nimble/host/src/ble_gap.c

```diff
@@ -3384,43 +3419,108 @@ ble_gap_periodic_adv_stop(uint8_t instance)
+    psync = ble_npl_event_get_arg(ev);
+    cb = psync->cb;
+    cb_arg = psync->cb_arg;
+
+    memset(&event, 0, sizeof event);
+
+    event.type = BLE_GAP_EVENT_PERIODIC_SYNC_LOST;
+    event.periodic_sync_lost.sync_handle = psync->sync_handle;
+    event.periodic_sync_lost.reason = BLE_HS_EDONE;
+
+    /* Free the memory occupied by psync as it is no longer needed */
+    ble_hs_periodic_sync_free(psync);
+
+    ble_gap_event_listener_call(&event);
+    if (cb) {
+        cb(&event, cb_arg);
+    }
+}
+
 int
-ble_gap_periodic_adv_create_sync(const ble_addr_t *peer_addr, uint8_t adv_sid,
-                                 uint8_t filter_policy, uint16_t skip,
-                                 uint16_t sync_timeout)
+ble_gap_periodic_adv_create_sync(const ble_addr_t *addr, uint8_t adv_sid,
+                                 const struct ble_gap_periodic_sync_params *params,
+                                 ble_gap_event_fn *cb, void *cb_arg)
 {
     uint8_t buf[BLE_HCI_LE_PERIODIC_ADV_CREATE_SYNC_LEN];
+    struct ble_hs_periodic_sync *psync;
     uint16_t opcode;
     int rc = 0;
 
-    if (!ble_hs_periodic_sync_can_alloc()) {
-        return BLE_HS_ENOMEM;
-    }
-    opcode = BLE_HCI_OP(BLE_HCI_OGF_LE,
-                                    BLE_HCI_OCF_LE_PERIODIC_ADV_CREATE_SYNC);
+    ble_hs_lock();
 
     /* No sync can be created if another sync is still pending */
-    if (ble_gap_master.pending_create_sync) {
+    if (ble_gap_sync.op == BLE_GAP_OP_SYNC) {
+        ble_hs_unlock();
         return BLE_HS_EBUSY;
     }
 
-    rc = ble_hs_hci_cmd_build_le_periodic_adv_create_sync(filter_policy,
-                                                          adv_sid,
-                                                          peer_addr->type,
-                                                          peer_addr->val,
-                                                          skip, sync_timeout,
-                                                          buf, sizeof(buf));
+    /* cannot create another sync if already synchronized */
+    if (ble_hs_periodic_sync_find(addr, adv_sid)) {
+        ble_hs_unlock();
+        return BLE_HS_EALREADY;
+    }
+
+    /* preallocate sync element */
+    psync = ble_hs_periodic_sync_alloc();
+    if (!psync) {
+        ble_hs_unlock();
+        return BLE_HS_ENOMEM;
+    }
+
+    ble_npl_event_init(&psync->lost_ev, ble_gap_npl_sync_lost, psync);
+
+    if (addr) {
+        rc = ble_hs_hci_cmd_build_le_periodic_adv_create_sync(0x00,
+                                                              adv_sid,
+                                                              addr->type,
+                                                              addr->val,
+                                                              params->skip,
+                                                              params->sync_timeout,
+                                                              buf, sizeof(buf));
```

Comment:
tbh in such case I'd just indent by 4 spaces instead of aligning with opening parenthesis :)